crates/currency/src/lib.rs

```diff
@@ -14,23 +15,419 @@ use frame_support::{
+            let signed_inner =
+                TryInto::<SignedInner<T>>::try_into(self.amount).map_err(|_| Error::<T>::TryIntoIntError)?;
+            let signed_fixed_point = <T as pallet::Config>::SignedFixedPoint::checked_from_integer(signed_inner)
+                .ok_or(Error::<T>::TryIntoIntError)?;
+            Ok(signed_fixed_point)
+        }
+
+        pub fn min(&self, other: &Self) -> Result<Self, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(if self.le(other)? { self.clone() } else { other.clone() })
+        }
+
+        pub fn lt(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount < other.amount)
+        }
+
+        pub fn le(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount <= other.amount)
+        }
+
+        pub fn eq(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount == other.amount)
+        }
+
+        pub fn ge(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount >= other.amount)
+        }
+
+        pub fn gt(&self, other: &Self) -> Result<bool, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            Ok(self.amount > other.amount)
+        }
+
+        // todo: all other checked_ math operations
+
+        pub fn transfer(&self, source: &T::AccountId, destination: &T::AccountId) -> Result<(), DispatchError> {
+            with_currency_id::transfer::<T>(self.currency_id, source, destination, self.amount)
+        }
+
+        pub fn lock(&self, account_id: &T::AccountId) -> Result<(), DispatchError> {
+            with_currency_id::lock::<T>(self.currency_id, account_id, self.amount)
+        }
+
+        pub fn unlock(&self, account_id: &T::AccountId) -> Result<(), DispatchError> {
+            with_currency_id::unlock::<T>(self.currency_id, account_id, self.amount)
+        }
+
+        pub fn burn(&self, account_id: &T::AccountId) -> DispatchResult {
+            with_currency_id::burn::<T>(self.currency_id, account_id, self.amount)
+        }
+
+        pub fn mint(&self, account_id: &T::AccountId) -> DispatchResult {
+            with_currency_id::mint::<T>(self.currency_id, account_id, self.amount)
+        }
+
+        // lock, unlock, etc
+
+        pub fn convert_to(&self, currency_id: CurrencyId<T>) -> Result<Self, DispatchError> {
+            T::CurrencyConversion::convert(self, currency_id)
+        }
+
+        pub fn is_zero(&self) -> bool {
```

Comment:
We should implement `Zero` in `num-traits`.

Reply:
`Zero` requires a `fn zero() -> Self` implementation, which due to the `currency_id` does not really make sense